File: ModoMercado/Core/Domain/Models/ItemDeLista.swift
import Foundation

struct ItemDeLista: Identifiable, Equatable, Codable {
    let id: UUID
    let produtoId: UUID
    var quantidade: Double
    var concluido: Bool
    var ultimoPrecoRegistrado: Decimal?   // snapshot no momento da adição

    var subtotalEstimado: Decimal {
        guard let preco = ultimoPrecoRegistrado else { return .zero }
        return preco * Decimal(quantidade)
    }

    init(
        id: UUID = UUID(),
        produtoId: UUID,
        quantidade: Double,
        concluido: Bool = false,
        ultimoPrecoRegistrado: Decimal? = nil
    ) {
        self.id = id
        self.produtoId = produtoId
        self.quantidade = quantidade
        self.concluido = concluido
        self.ultimoPrecoRegistrado = ultimoPrecoRegistrado
    }
}

File: modo_mercado/lib/core/domain/models/item_de_lista.dart
import 'package:equatable/equatable.dart';

class ItemDeLista extends Equatable {
  final String id;
  final String produtoId;
  final double quantidade;
  final bool concluido;
  final int? ultimoPrecoRegistradoCentavos;

  const ItemDeLista({
    required this.id,
    required this.produtoId,
    required this.quantidade,
    required this.concluido,
    this.ultimoPrecoRegistradoCentavos,
  });

  int get subtotalEstimadoCentavos {
    final preco = ultimoPrecoRegistradoCentavos;
    if (preco == null) return 0;
    return (preco * quantidade).round();
  }

  ItemDeLista copyWith({
    String? id,
    String? produtoId,
    double? quantidade,
    bool? concluido,
    int? ultimoPrecoRegistradoCentavos,
  }) {
    return ItemDeLista(
      id: id ?? this.id,
      produtoId: produtoId ?? this.produtoId,
      quantidade: quantidade ?? this.quantidade,
      concluido: concluido ?? this.concluido,
      ultimoPrecoRegistradoCentavos:
          ultimoPrecoRegistradoCentavos ?? this.ultimoPrecoRegistradoCentavos,
    );
  }

  @override
  List<Object?> get props => [
        id, produtoId, quantidade, concluido,
        ultimoPrecoRegistradoCentavos,
      ];
}
